### ISSUE TITLE
Improve error message when missing `queryGenerationFactories` in the template.

### ISSUE THREAD

**Issue by efstathiosntonas** [magidoc-org/magidoc#92]
Thank you for this great generator.

While Build Template it throws this error:

```
 Error: Cannot generate a random value for scalar 'smallint'.
  The random generator is not able to randomly generate a value for non-standard GraphQL scalars.
  You have to provide a custom factory by providing this in your config:
  {
  'smallint': () => generateRandomCustomScalar()
  }
```

Can you please provide an example on how to fix this? Documentation is not clear enough, where the ` 'smallint': () => generateRandomCustomScalar()` should go in the config? Can you provide an example of `generateRandomCustomScalar()`?

**Comment by pelletier197** [comment]
Thank you for raising this issue. So this is an error thrown by the [Query generator plugin](https://magidoc-org.github.io/magidoc/plugins/graphql-query-generator), which is not entirely accurate when using the magidoc template directly. I will improve the error message for this special error, since I expect it may happen quite often. 

To solve it, you need to provide a `query generation factory` in the magidoc config 

```javascript
export default {
  introspection: {
    // ...
  }, 
  website: {
    // ...
    queryGenerationFactories: {
      smallint: 42
    }
  }
}
```

More detail about it can be found [here](https://magidoc-org.github.io/magidoc/plugins/graphql-query-generator].

**Comment by pelletier197** [comment]
Oh.. sorry, my bad. I wrote this on my phone, should have opened my computer 😅. Simply wrap it in an `options` key:

```javascript
export default {
  introspection: {
    type: "url",
    url: "http://localhost:9091/v1/graphql",
    headers: {
      "x-hasura-admin-secret": "myadminsecretkey",
    },
  },
  website: {
    template: "carbon-multi-page",
    options: {
      queryGenerationFactories: {
        smallint: 42
      }
    }
  },
};
```

**Comment by efstathiosntonas** [comment]
yeah, you are right, on the docs link you provided above it's inside an `options` object but was hard to spot due to the many comments 😅.

Probably this is not possible: Can we first check for unknown scalars prior reaching the Build Template step? It takes lot's of time to build and throw an error if scalar is not known. We know witch scalars are not compatible before hand of course, just saying.

Thanks @pelletier197, keep it up!

**Comment by pelletier197** [comment]
I can try to look if I could find a way to do that, or at least show all the missing scalars at once in the error. 

You're right, it's not necessarily easy, because the cli doesn't actually do anything with the `options`, it simply forwards them directly to the template. And since the factories support glob syntax as well, it adds a little bit of complexity. 

I'll see what I can do, thanks for the idea!

**Comment by efstathiosntonas** [comment]
That makes sense!

 It would be great to do a check first, will save some time while building the config since we won't have to cross check our scalars against it. 

Another option is to have them predefined, at least the most common ones like eg `timestamp`, `date`, `uuid` and let the user override them on `options.queryGenerationFactories` if desired.

**Comment by pelletier197** [comment]
There are some [default factories](https://github.com/magidoc-org/magidoc/blob/main/packages/plugins/query-generator/src/generator/defaultFactories.ts), but very few. I'll add those you mentioned, they clearly seem like common things that could come back in many APIs.

**Comment by efstathiosntonas** [comment]
great, these are some common Hasura scalars:

```json
    {
        _text: "loremipsum",
        bigint: 1000,
        date: "2022-03-06T08:23:45.000Z",
        float8: 0.1,
        geography: { type: "Point", coordinates: [0, 0] },
        geometry: { type: "Point", coordinates: [0, 0] },
        jsonb: {},
        numeric: 1,
        smallint: 42,
        timestamp: "2022-03-06T08:23:45.000Z",
        timestamptz: "2022-03-06T08:23:45.000Z",
        uuid: "a12ce80c-42e3-4149-89cc-4afda30e5e7f",
      }
```

After adding all these scalars above it throws a new error:

```bash
  Error: Fatal error: A recursive input with a non-nullable circular reference was detected, which makes it impossible to generate a query. If you ever see this error, it means that your GraphQL uses an invalid schema.
  at generateInput (file:///private/var/folders/yz/dk5dbx8n2hqcfk1kdktxmknh0000gn/T/carbon-multi-page@2.0.0/node_modules/.pnpm/@magidoc+plugin-query-generator@2.0.0/node_modules/@magidoc/plugin-query-generator/build/generator/fakeGenerator.js:34:23)
```

ps. I'm using Hasura 2.7.0

**Comment by pelletier197** [comment]
Thanks. Those all seem reasonable to add to the default factories. 

About your error, thanks for taking the time to check it. 

I'll check it out as well when I get some time. Would it be possible for you to provide me with the `_schema.json` file for your API? If not possible, a minimal reproduction schema would be great otherwise. 

I feel like this error message needs more detail as well. I'll improve that as well.

**Comment by efstathiosntonas** [comment]
Sorry but I cannot share anything, it's a private repo. :(

It fails on the `account_insert_input`, this is how I found it:

```js
 if (!isNullableType(input)) {
        console.log(
          "type: ",
          unwrappedType,
          "input: ",
          input,
          "name: ",
          unwrappedType.name
        );
        throw new Error(
          "Fatal error: A recursive input with a non-nullable circular reference was detected, which makes it impossible to generate a query. If you ever see this error, it means that your GraphQL uses an invalid schema."
        );
      }

```

this spits out `type:  "account_insert_input" input:  "account_insert_input!" name:  account_insert_input`, this input type has 0 non-null types and this table has some (ORM) table connections: eg:

```
input account_insert_input {
 about_me: String,
show_seen_sent: Boolean
    targeted_matches: match_arr_rel_insert_input
    time_format: String
    unban_reason: String
    unbanned_at: timestamptz
    unbanned_by: account_obj_rel_insert_input
    unbanned_by_id: String
    updated_at: timestamptz
    verified_by: account_obj_rel_insert_input
    verified_by_id: String
devices: device_arr_rel_insert_input
```

where `device_arr_rel_insert_input` type is:

```
"input type for inserting array relation for remote table \"device\""
input device_arr_rel_insert_input {
    data: [device_insert_input!]!
    "upsert condition"
    on_conflict: device_on_conflict
}

```

`account_obj_rel_insert_input` type is:
```
"input type for inserting object relation for remote table \"account\""
input account_obj_rel_insert_input {
    data: account_insert_input!
    "upsert condition"
    on_conflict: account_on_conflict
}

```

wish I could provide more info, best thing you can do is to run a hasura docker and create a simple database. Hasura has weird schema in general.

**Comment by pelletier197** [comment]
Okay, I think I get the issue. I'll open another bug for that. This is recursion with two layers, I'm gonna see how I can fix this, not an easy one. 

In the meantime, you could probably do something like this if you're blocked and it's the only thing like this in your API.

```javascript
{
  queryGenerationFactories: {
    device_arr_rel_insert_input: null,
    'device_arr_rel_insert_input!': {} // an example
   }
}
```

Thanks for all the help by the way, really appreciated!

**Comment by efstathiosntonas** [comment]
You're welcome @pelletier197, and again, thanks for this great generator. If you need anything please tag me! If you want to get exactly the same experience like me then visit [Hasura Getting Started](https://hasura.io/docs/latest/graphql/core/getting-started/docker-simple/).

**Comment by pelletier197** [comment]
The issue with the recursion should be fixed now in 2.1.0, along with a few other improvements. Let me know if you have other issues. 

I have not implemented the default types of Hasura yet, cause I have a little bit of confusion looking at the types example you shared compared to Hasura's docs.  I'll need a little more research before implementing them. See #93 for more details.
